sdk/inc/azure/core/az_json.h

```diff
@@ -80,6 +76,13 @@ typedef struct
   /// In the case of JSON strings, the slice does not include the surrounding quotes.
   az_span slice;
 
+  // Avoid using enum as the first field within structs, to allow for { 0 } initialization.
+  // This is a workaround for IAR compiler warning [Pe188]: enumerated type mixed with another type.
+
+  /// This read-only field gives access to the type of the token returned by the #az_json_reader,
+  /// and it shouldn't be modified by the caller.
+  az_json_token_kind kind;
+
```

Comment:
I want to learn more about packing and padding. I definitely see why it's better where you moved the enum (1 byte?) than where it was, however, why not move it below the int32_t size?  The next data type is a struct that begins with two bools (2 bytes total?).  Would the compiler see the enum next to the two bools and only pad 1 more byte? or is there a cut-off at the struct, so it doesn't matter?  This is probably easier to discuss in a call.

Reply:
Sure, we can chat. I believe enums in C are `int`-backed, so 4 bytes.

To clarify, I didn't re-order them for maximum struct packing, but we could.

Reply:
When initializing a struct, the first initializer in the list initializes the first declared member and omitting fields in initialization  implicitly initialized the same as for objects that have static storage duration.

This above principle is getting used here to initialize the struct with zero, as you can see in line azure-sdk-for-c/sdk/src/azure/core/az_json_writer.c:774. 

Now, in this (az_json_writer.c:774) initialization we are initializing enum with 0, which is causing issue in compilers (like IAR) that do not want INT to be mixed with enum. Now with shuffling the fields, we are avoiding the warning.